api/agentpb/agent.proto

```diff
@@ -81,6 +81,48 @@ message SetStateResponse {
 
 }
 
+// ActionName defines allowed action names.
+enum ActionName {
+  ACTION_NAME_INVALID = 0;
+  PT_SUMMARY = 1;
+  PT_MYSQL_SUMMARY = 2;
+  MYSQL_EXPLAIN = 3;
+}
+
+// StartActionRequest represents request to run an action by name.
+message StartActionRequest {
+  string action_id = 1;
+  ActionName name = 2;
+  repeated string parameters = 3;
```

Comment:
is only slice of strings enough to run any action?

Reply:
We have discussed this a little with @AlekSi and @rnovikovP. But I'm not sure. I think at this moment it's enough, at least for all actions running in the `shell`. It can be something like:

```golang
[]string{"-flag1","value_of_flag1","-flag2","-flag3","value_of_flag3"}
```

The same thing, in theory, can be used for other actions. If we need `K=>V`, we can use `odd`
elements for `Keys` and `even` for `Values`. For example:
```golang
[]string{"key1","value1","key2","value2"}
```

Reply:
By that logic, we can also push action_id and type into nameless strings:
```go
[]string{"Action_id","pt-summary","key1","value1","key2","value2"}
```
But that hardly a good API.

Let's use common parameters for all process actions and type-specific parameters for other action. Something like:
```proto
message ProcessActionParams {
  repeated string args = 1;
}

message StartActionRequest {
  string action_id = 1;
  ActionType type = 2;
  oneof params {
    ProcessActionParams process_action_params = 3;
    MySQLExplainActionParams …
    MySQLShowCreateTableParams …
  }
}
```

I will have more comments, but that's something we should start changing now.

Reply:
I agree except the concept of `ProcessActionParams` it looks bad as we broke encapsulation and give the user some knowledge about internal action nature, as some actions, in fact, are processes. It's also a bad API I think.

Reply:
Do you suggest to make a separate `params` message for each action type? even if they are the same?

Reply:
> Do you suggest to make a separate `params` message for each action type? even if they are the same?

Yep, maybe. Also, we can use unique params for all actions, or some generalized structure for all actons. Maybe map for example.

In fact, they are no the same (I mean nevermind `pt-summary` and `pt-mysql-summary` is shell scripts. They have different parameters, they also have different meanings). They will the same only if you know that they are `shell scripts` but this breaks encapsulation at once.

If we made an abstraction, we should hide the implementation details. If we show implementation details, abstraction is useless.

Reply:
I agree that it should be unique params for all actions. So it can help us to not let users do whatever they want.

Reply:
That's not a user-visible API. Arguments about encapsulation miss that point.
Let's also not redefine the scope of that task, and definition of action ATM.

> I agree that it should be unique params for all actions.

My only concern there that all process actions are not, in fact, different at that level (pmm-managed<->pmm-agent protocol). All process actions parameters will have the same structure. We will just add more types without benefits.

Reply:
> Let's also not redefine the scope of that task, and definition of action ATM.

The scope of this task isn't clean, so this task is discussable I think.

> My only concern there that all process actions are not, in fact, different at that level (pmm-managed<->pmm-agent protocol)

So, if we know about internal actions structure at this level we shouldn't use actions at all here. In this case, this abstraction need only in `pmm-managed<->UI`? In this case, I think we should rethink an API between `pmm-managed<->pmm-agent`. As I said above we probably should have an API like `RunShellScript, RunSQLQuery, RunMongoDBCommand` in that level, why not?

Reply:
One more thing. This API does not interfere to merge all this stuff as we haven't other actions like `EXPLAIN` and we have a clean separation in code between `pmm-managed<->UI` and `pmm-managed<->pmm-agent`. So we can merge this and then rethink and refactor `pmm-managed<->pmm-agent` side without doing changes in Public API.

Reply:
> Also when I say user, I always mean person/program who call this API.

I'm sorry, but you said "end-user", and that's not what it means.

> As I said above we probably should have an internal API like RunShellScript, RunSQLQuery, RunMongoDBCommand in that level, why not?

We can't due to our security model: we should not provide a way to run arbitrary commands or SQL queries.
Again, let's not redefine the scope. It was clear last time we talked about.

There is no way for `repeated string parameters` for all Actions to be merged. That's my call.

Reply:
> I'm sorry, but you said "end-user", and that's not what it means.
Ok. I'll be more accurate next time and will separate user and `end-user`. 
In this case, I mean `person/program`.

> We can't due to our security model: we should not provide a way to run arbitrary commands or SQL queries.

We can if commands will be predefined, not arbitrary. We are talking about clean API and good separation not about having arbitrary commands and queries.

All I'm trying to say is that suggested separation on `ProcessActionParams` and `AnythingElseActionParams` looks awkward and also breaks encapsulation. 

In fact, all MySQL parameters can be represented as a `map`, or as a string slice, like in SQL libraries. Why shell commands are so different and special? Why we can group process actions` but can't group `MySQL query actions` for example? I think we could group it better.

> There is no way for repeated string parameters for all Actions to be merged. That's my call.

I agree.